base/regex.jl

```diff
@@ -212,14 +212,30 @@ julia> hr
 "11"
 ```
 """
-struct RegexMatch <: AbstractMatch
-    match::SubString{String}
-    captures::Vector{Union{Nothing,SubString{String}}}
+struct RegexMatch{S<:AbstractString} <: AbstractMatch
+    match::SubString{S}
+    captures::Vector{Union{Nothing,SubString{S}}}
     offset::Int
     offsets::Vector{Int}
     regex::Regex
 end
 
+RegexMatch(match::SubString{S}, captures::Vector{Union{Nothing,SubString{S}}},
+           offset::Union{Int, UInt}, offsets::Vector{Int}, regex::Regex) where {S<:AbstractString} =
+    RegexMatch{S}(match, captures, offset, offsets, regex)
+
+function RegexMatch{StyledString{S}}(m::RegexMatch{S}, str::StyledString{S}) where {S<:AbstractString}
+    RegexMatch{StyledString{S}}(
+        (@inbounds SubString{StyledString{S}}(
```

Comment:
If this method of `RegexMatch` is public, then, IIUC, this `@inbounds` allows folks to create a substring that violates invariants using only safe public API if `m` and `str` come from different strings:

```julia
julia> str = StyledString{String}("abc",[])
"abc"

julia> m = match(r"def", "abcdef")
RegexMatch("def")

julia> rm = RegexMatch{StyledString{String}}(m, str); # Should this error?

julia> m2 = rm.match;

julia> dump(m2)
SubString{StyledString{String}}
  string: StyledString{String}
    string: String "abc"
    properties: Array{Tuple{UnitRange{Int64}, Pair{Symbol, Any}}}((0,))
  offset: Int64 3
  ncodeunits: Int64 3
```

If it's a private method, then ideally it would have a name that isn't exported. I prefer the solution of renaming this to `_RegexMatch`. In general, I don't like adding private methods to public functions.

Reply:
Thanks @LilithHafner. This is just needed for the two `match` functions. I implemented this as a constructor because it, well, constructs a `RegexMatch` from an existing one.

I've just pushed this change:

```julia
function _styledmatch(m::RegexMatch{S}, str::StyledString{S}) where {S<:AbstractString}
    RegexMatch{StyledString{S}}(
        (@inbounds SubString{StyledString{S}}(
            str, m.match.offset, m.match.ncodeunits, Val(:noshift))),
        Union{Nothing,SubString{StyledString{S}}}[
            if !isnothing(cap)
                (@inbounds SubString{StyledString{S}}(
                    str, cap.offset, cap.ncodeunits, Val(:noshift)))
            end for cap in m.captures],
        m.offset, m.offsets, m.regex)
end

match(re::Regex, str::StyledString{S}) where {S<:AbstractString}
    m = match(re, str.string)
    if !isnothing(m)
        _styledmatch(m, str)
    end
end

function match(re::Regex, str::StyledString{S}, idx::Integer, add_opts::UInt32=UInt32(0)) where {S<:AbstractString}
    m = match(re, str.string, idx, add_opts)
    if !isnothing(m)
        _styledmatch(m, str)
    end
end
```

Even though it's no longer used the `{S<:AbstractString}` in the `match` methods seems to be needed to beat out the "convert to `String` and try again" fallback method in terms of method specificity.

Reply:
Lovely! 

> the {S<:AbstractString} in the match methods seems to be needed to beat out the "convert to String and try again" fallback method in terms of method specificity.

I don't understand how this is the case. IIUC, the set of alternative methods is
```
 [1] match(re::Regex, str::Union{SubString{String}, String}, idx::Integer)
 [2] match(r::Regex, s::AbstractString, i::Integer)
 [3] match(re::Regex, str::StyledString{S}) where S<:AbstractString
 [4] match(r::Regex, s::AbstractString)
 [5] match(re::Regex, str::Union{SubString{String}, String}, idx::Integer, add_opts::UInt32)
```
none of which are ambiguous with or more specific than
```
match(re::Regex, str::StyledString, idx::Integer, add_opts::UInt32=UInt32(0))
```

Reply:
Neither can I! When I removed them though, I saw:

```
  ArgumentError: regex matching is only available for the String type; use String(s) to convert
  Stacktrace:
   [1] match(r::Regex, s::StyledString{String}, i::Int64)
     @ Base ./regex.jl:458 
   [2] match(r::Regex, s::StyledString{String})
     @ Base ./regex.jl:457
```

where the methods [2] and [1] are:

```julia
match(r::Regex, s::AbstractString) = match(r, s, firstindex(s))
match(r::Regex, s::AbstractString, i::Integer) = throw(ArgumentError(
    "regex matching is only available for the String type; use String(s) to convert"
))
```

Thinking about it, the type restriction was the only thing that changed in the `match` method ... so I just added them back :shrug:.

Reply:
Ah, I've just recalled that the `julia` process I had there may have been from a version that tried scoping `_styledmatch`, which would mess things up :sweat_smile:.

I'll try re-checking and push the version we both expect would work if it actually does.